### ISSUE TITLE
Performance of some instructions could be significantly improved

### ISSUE THREAD

**Issue by Fish-Git** [SDL-Hercules-390/hyperion#101]
A quick review of a some instructions reveals they are very poorly coded, performance-wise. Most all of them have not been modified since the repository was initially setup over 17 years ago and use `vfetchb` and `vstoreb` to process their operands one byte at a time (which is _extremely_ inefficient!):

_**Note:** Instructions which are checked have been recently modified and now no longer need to be improved. The ones that are unchecked are those whose performance still needs to be improved._

- [x] &nbsp;&nbsp; `B241`  `CKSM`&nbsp;&nbsp;   Checksum
- [x] &nbsp;&nbsp; `A9`&nbsp;&nbsp;&nbsp; `CLCLE`  &nbsp;Compare Logical Long Extended &nbsp; _(contributed by James Wekel)_
- [x] &nbsp;&nbsp; `B25D`  `CLST`&nbsp;&nbsp;   Compare Logical String &nbsp; _(completed by Bob Polmanter)_
- [x] &nbsp;&nbsp; `B257`  `CUSE`&nbsp;&nbsp;   Compare Until Substring Equal &nbsp; _(contributed by James Wekel)_
- [x] &nbsp;&nbsp; `B2A7`  `CU12`&nbsp;&nbsp;   Convert UTF-8 to Unicode
- [x] &nbsp;&nbsp; `B9B0`  `CU14`&nbsp;&nbsp;   Convert UTF-8 to UTF-32
- [x] &nbsp;&nbsp; `B255`  `MVST`&nbsp;&nbsp;   Move String &nbsp; _(completed by Bob Polmanter)_
- [x] &nbsp;&nbsp; `B25E`  `SRST`&nbsp;&nbsp;  Search String &nbsp; _(completed by Bob Polmanter)_
- [x] &nbsp;&nbsp; `B2A5`  `TRE`&nbsp;&nbsp;&nbsp;&nbsp;    Translate Extended &nbsp; _(contributed by James Wekel)_
- [x] &nbsp;&nbsp; `B9BF`  `TRTE`&nbsp;&nbsp;   Translate and Test Extended &nbsp; _(contributed by James Wekel)_
- [x] &nbsp;&nbsp; `D0`&nbsp;&nbsp;&nbsp;&nbsp;    `TRTR`&nbsp;&nbsp;   Translate and Test Reverse &nbsp; _(contributed by James Wekel)_
- [x] &nbsp;&nbsp; `B9BD`  `TRTRE`  Translate and Test Reverse Extended &nbsp; _(contributed by James Wekel)_

I believe that by using the same or similar techniques as were used to improve the performance of the CLCL, MVCIN and TRT instructions (see issue [#99: Poor performance of CLCL, MVCIN and TRT instructions](https://github.com/SDL-Hercules-390/hyperion/issues/99)), the performance of each of them could be _**significantly**_ improved (_very likely by at least one, and possibly two, orders of magnitude!_).

I would like to invite any/all Hercules developers to _please_ try their hand at doing so.

_I could use some help here!_

This is a low priority item, but really should eventually be looked into. While I have not performed any measurements in this area, I rather suspect some of these instructions(*) might well be used quite extensively.

Bottom line: there's _lots_ of room for improvement here!
___
`(*)` The MVST (Move String) SRST (Search String) instructions appears to have been specifically written to handle the C/C++ `strcpy` and `strchr` library functions for example, and TRTE is a perfect candidate for the C/C++ `memchr` library function, etc.

**Comment by Fish-Git** [comment]
> Since @ivan-w has volunteered in Issue #108 to look into the TRTR, TRTE and TRTRE (and hopefully TRE too!) instructions, I am marking this issue as "In Progress...".
> 
> Apparently(*) however, I cannot assign this issue to him until he first posts a comment to this GitHub Issue. Once he does then I'll do that.
> 
> (*) The only choices GitHub appears to offer for assignment are individuals participating in a given issue's conversation (i.e. only those that choose to reply). Go figure.

The above is no longer true now that @ivan-w is now a member of our organization, so I've gone ahead and assigned this issue to Ivan.

Ivan? If this is unacceptable, then please feel free to remove yourself as the assignee!

**Comment by wably** [comment]
@Fish-Git I think I will sign up to do the three string instructions on the list: CLST, MVST, and SRST.  Unless you know that someone else is working on them?

All three of these could benefit from using techniques similar to those already used for some other instructions mentioned, making allowances for access exceptions in the right order and considerations for page boundaries.

**Comment by Fish-Git** [comment]
> I think I will sign up to do the three string instructions on the list: CLST, MVST, and SRST. Unless you know that someone else is working on them?

Go for it! _PLEASE!_&nbsp;`:)`

Ivan volunteered (or implied he volunteered) in another issue (#108), but I have no idea whether or not he is actually working on it. You should probably coordinate your effort with him off-list. He may just be busy with work or real life. (whatever the heck those things are!)

Thanks!

**Comment by wably** [comment]
Since nothing seemed to be happening here I went ahead and completed improvements to the three string instructions, MVST, SRST, and CLST.  All three had their execution times reduced by over 80%.  Below summarizes the results:

                 before     after      reduction %
                 ----------------------------------
        MVST     9.62       1.21         87%
        SRST     5.40       1.01         81%
        CLST     10.05      1.67         83%

All three timings were moves, searches, or compares across three page frames, with 100,000 iterations each.  All times are in seconds. The host computer was an Intel i7.  SRST 'before' results are approximately one-half of the others because only one operand has to be incremented.  The execution times were improved by implementing the page boundary dance for each operand thereby avoiding a vfetch( ) or vstore( ) for every byte.

I will commit this code in a day or two.  I just want to double check everything and ensure that I have all of the test cases covering all condition codes and operand combinations (crossing pages, not crossing, etc).

**Comment by Fish-Git** [comment]
> Since nothing seemed to be happening here I went ahead and completed improvements to the three string instructions, MVST, SRST, and CLST. All three had their execution times reduced by over 80%.

Wow! &nbsp; Fantastic work, Bob! &nbsp; _Thank you!_

> I will commit this code in a day or two. I just want to double check everything and ensure that I have all of the test cases covering all condition codes and operand combinations (crossing pages, not crossing, etc).

I look forward to seeing your commit!&nbsp;`:)`

**Comment by wably** [comment]
String instructions committed by 8194b34

@Fish-Git if you are still seeing problems with tabs in the source, please inform.

**Comment by Fish-Git** [comment]
> String instructions committed by [8194b34](https://github.com/SDL-Hercules-390/hyperion/commit/8194b34943280233ebbb33284880d3bdb0c13d42)

Thank you, Bob!  Have you created a `runtest` test case for these instructions yet?

> if you are still seeing problems with tabs in the source, please inform.

I checked and I didn't see any.

**Comment by wably** [comment]
Fish,

My test cases are simple assembler programs that run under VM (they could also run under MVS with some JCL wrapped around them).  There are three programs, one for each instruction improved.  Each program does a number of tests, basically it sets up a situation and then executes the instruction.  The results of the execution are displayed and it is compared to the known expectation (basically the results of these exact same tests before my modification). Then on to the next situation and test for the same instruction.  Each possible condition code and combinations of page crossing operands (or not) were exercised and checked against known results.

So after all of that and to answer your question, no I do not have a ‘runtest’ test case; I dont even know what that is.  I can make my programs available of course, if that is acceptable.

**Comment by srorso** [comment]
Hi Bob:

If you would post / PM the programs to me, I  would be happy to create runtest cases.  

Steve

**Comment by wably** [comment]
Steve,

Thanks for the offer, but you shouldn't have to do what I should have done in the first place.

After my response to Fish last night, I remembered that there was a 'tests' directory in the repo, and when I looked there I saw the kinds of tests that he was referring to.  So, I can go ahead and create tests similar to those as appropriate for the string instructions.

It will definitely be more of a pain to put these together than the ALC programs I have now, but I recognize some reasons for having the tests be more or less standardized and not reliant on IPLing an OS prior to running a test.  The learning curve to put these together effectively is the painful part; I can whip up stuff in ALC much faster and then move on.

Bob

**Comment by srorso** [comment]
Hi Bob,

I understand fully, but the offer stands.  Teamwork and all that.  

Harold Grovesteen's [SATK](https://github.com/s390guy/SATK) package is a worthy tool, and I am happy to field questions, particularly about the BFP examples.  

Steve

**Comment by wably** [comment]
This is definitely a case of “no good deed goes unpunished”.

**Comment by Fish-Git** [comment]
> This is definitely a case of "no good deed goes unpunished".

(LOL!) &nbsp;I just asked a simple question!&nbsp;`:)`

**Comment by wably** [comment]
Fish, I reviewed your changes and I am fine with them.  Now that I can see how you want things to look I can make a better effort to comply with it.  Actually, if you published a 'coding standards' document for that sort of thing, then it would be a lot easier to follow the guideline - if the look of the code is important to you.

**Comment by Fish-Git** [comment]
Here are the rules I try to live by (I don't always adhere to many of them but I usually try to for _most_ of them) (I think):

(https://www.goodreads.com/book/show/598624.Writing_Solid_Code)

<h1 align=center>Writing Solid Code</h1>

<h3 align=center>by Steve Maguire</h3>

* Always ask, "Can this variable or expression over- or underflow?"
* Always look for, and eliminate, flaws in your interfaces.
* As you step through code, focus on data flow.
* Create thorough subsystem checks, and use them often.
* Define explicit function arguments. Eliminate ambiguity.
* Design your tests carefully. Nothing should be arbitrary.
* Document unclear assertions.
* Don't allow unnecessary flexibility.
* Don't clean up code unless the cleanup is critical to the product's success.
* Don't fix bugs later; fix them now.
* Don't hide bugs when you program defensively.
* Don't implement nonstrategic features. There is no free lunch.
* Don't keep 'trying' solutions until you find one that works. Take the time to find the correct solution.
* Don't reference memory that you don't own.
* Don't wait until you have a bug to step through your code.
* Don't write multipurpose functions. Write separate functions to allow stronger argument validation.
* Either remove implicit assumptions, or assert that they are valid.
* Eliminate random behavior. Force bugs to be reproducible.
* Enable all compiler warnings.
* Fix the cause, not the symptom.
* Handle special cases just once.
* If something happens rarely, force it to happen often.
* Implement your designs as accurately as possible.
* Maintain both ship and debug versions of your code.
* Make code intelligible at the point of call. Avoid Boolean arguments.
* Make it hard to ignore error conditions. Don't bury error codes in return values.
* Never allow the same bug to bite you twice.
* Shred your garbage.
* Step through every code path.
* Strip undefined behavior from your code.
* Strive to implement transparent integrity checks.
* Throw away your bag of tricks. Be truly clever: Write boring code.
* Tight C code does not guarantee efficient machine code.
* Use a second algorithm to validate your results.
* Use assertions to detect impossible conditions.
* Use assertions to validate all function arguments.
* Use lint to catch bugs your compiler may miss.
* Write and test code in small chunks. Always test your code, even if that means your schedule will slip.
* Write comments that emphasize potential hazards.
* Write functions that, given valid inputs, cannot fail.

**Comment by wably** [comment]
Runtest cases for the three improved string instructions added by commit 44e4d07

Many thanks to Steve Orso, who provided assistance, examples, tips and advice in getting started with runtest cases!

**Comment by Fish-Git** [comment]
> Runtest cases for the three improved string instructions added by commit [44e4d07](https://github.com/SDL-Hercules-390/hyperion/commit/44e4d078c0f8d192b9e02b7bf75b04befc47c4db)
> 
> Many thanks to Steve Orso, who provided assistance, examples, tips and advice in getting started with runtest cases!

I appreciate the effort, guys! THANKS!&nbsp;`:)`

As new instructions are added (or existing ones enhanced) I feel it's important to add a runtest test case to our existing suite that verifies the instruction is [still] working properly. So I really appreciate the extra effort, guys. Thanks!
___
And just as an FYI, the proper way to "restore" an `ostailor` setting is to use `ostailor default`, not null.

That is to say, `ostailor null` is an ostailor setting unto itself that completely disables the displaying of _any_ program interrupts (suppresses them all), but it is _**not**_ the _default_ setting when Hercules first comes up. That's what `ostailor default` is for: it's sets the _**default**_ ostailor value.

It's no big deal. I've already fixed it for you. It's just an FYI.

**Comment by wably** [comment]
@Fish-Git or @ivan-w do you know if the Translate instructions in the list at the beginning of this issue can be checked off as completed, by the changes made in issue #107 and #108   ?

**Comment by ivan-w** [comment]
The Translate and Test (and various variations) are complicated to test for compliance, but it's certainly feasible. The specific issue with Translate and Test instructions is that any access exception only occurs 'as needed' - so if the 2nd operand crosses a page boundary (and the 2nd page isn't accessible) but the data in the 1st operand doesn't mandate access the second page of the second operand, then everything should work without any exception. A complete unit test to test all conditions is tricky !

**Comment by wably** [comment]
@ivan-w right, that I understand.  But back to the question...do those changes you made also address the performance issue or is that still to be tackled?

**Comment by ivan-w** [comment]
The performance issues have been dealt with. TRT only required a few adjustment to deal with boundary issues just to be architecture compliant ! The original version went through translation for every step... The optimized version did translation as a block transaction - but would issue an unnecessary exception in some borderline case. The current version has the optimized version and deals with the borderline case. The current issue is the test case which should test all the conditions for possible exception (or lack thereof)

**Comment by wably** [comment]
Improved performance for the CKSM instruction as well as runtest cases added by commit 9f27856

The execution time for CKSM was reduced by 82%, from 10.98 seconds to 1.89 seconds  (500,000 iterations of CKSM on a 14,000 byte buffer of mixed data).

**Comment by Fish-Git** [comment]
> The COBOL program shown below takes about 50 seconds to run on Hercules. This is compared with 6 seconds when run on a DEC PDP-10 36 bit Tops-10 using the klh10 emulator and DEC's COBOL compiler of the same exact COBOL benchmark program.

Immaterial. IBM mainframes use a completely different machine architecture than the DEC PDP-10. They are not the same (nor even remotely similar). The two machine architectures are completely different from one another, and as a result, so are the hardware emulators themselves too. You are comparing apples to oranges.

> The MVS 3.8J COBOL benchmark program I've been using is shown below. Since the variable Y is being displayed, it forces the compiler to do all the intermediate calculations. _**The use of the PMAP option shows the assembler object code generated.**_
> 
> * [benchcob.txt](https://github.com/SDL-Hercules-390/hyperion/files/8765664/benchcob.txt)

_(emphasis added)_

You are presuming that we too are also familiar with, and run, the IBM MVS 3.8J operating system on Hercules. Such a presumption is erroneous. I know nothing about IBM MVS 3.8J, so your providing the COBOL source code is personally unhelpful to me.

What would be more helpful would be seeing the actual object code (i.e. the actual hardware instructions) that the IBM MVS 3.8J COBOL compiler generates (i.e. this presumed "PMAP" output you mention) so that we can review each of the instructions involved to see whether there is maybe anything we can do to improve their respective performance on Hercules.

Thanks.

**Comment by wrljet** [comment]
Fish,

I forget the exact date, but the *legal* COBOL compiler most people use with MVS 3.8, dates from around 1970 or something.
Came from the OS/360 MVT era and has been patched to run with modern dates, if I remember correctly.
(I'm not a COBOL guy)

Bill

**Comment by Fish-Git** [comment]
> Fish,
> 
> I forget the exact date, but ...

And this means ..................... what? Are you trying to explain why it doesn't (or likely doesn't) have any type of optimization option? It's unclear to me what information your comment is trying to convey.

**Comment by wgs777** [comment]
I've attached the complete untruncated listing.

[listing.txt](https://github.com/SDL-Hercules-390/hyperion/files/8773796/listing.txt)

We can see it is calling on this function: ILBOBID2
L     15,00C(0,12)          V(ILBOBID2)

Searching for ILBOBID2 in the IBM Compiler manual for this compiler "CB545 V2 LVL78 01MAY72"
http://www.bitsavers.org/pdf/ibm/360/os/cobol/GC28-6399-2_COBOL_Compiler_and_Library_Version_2_Programmers_Guide_Jul72.pdf

shows its a COBOL Library Conversion subroute for converting from Binary to Internal Decimal (browser is slow with find commands in this PDF, it initially shows not found, but if patient, it will then populate with the found search in the pdf).

So you may be correct, that the compiler is generating inefficient assembler code using decimal numbers and inefficient library routines.  Unfortunately we are stuck with this compiler as it's the only legal IBM Cobol Compiler available for TK4-.

I'm planning to do a large amount of Cobol development using TK4- and this compiler.  It would be beneficial to check the Hercules implementations of the assembler instructions that appear most often in the attached compiler listing (like you suggested): (ZAP/PACK/UNPK/AP/SP) to look for any opportunities to improve/optimize the Hercules code.

IBM hardware historically been used primarily for business programs, and most business programs are written in Cobol.  Optimizing Hercules' instruction implementations for assembler instructions that are commonly generated by Cobol compilers could really be a worthwhile improvement to the Hercules emulator.

**Comment by wrljet** [comment]
> > Fish,
> > I forget the exact date, but ...
> 
> And this means ..................... what? Are you trying to explain why it doesn't (or likely doesn't) have any type of optimization option? It's unclear to me what information your comment is trying to convey!

I'm suggesting it may not have the best optimization.

**Comment by Fish-Git** [comment]
> So you are correct. The compiler appears to be generating inefficient assembler code using decimal numbers and inefficient library routines. I checked the manual and this compiler doesn't have an optimize option. Unfortunately we are stuck with this compiler as it's the only legal IBM COBOL Compiler available for TK4-.

It's been far too many years since I've messed with COBOL, but have you investigated making changes to your program itself to be more efficient? For example, you have variables A, B and C defined as "COMP" (i.e. Binary?), but variables X and Y are both defaulting to non-binary (i.e. Decimal?). Since your `COMPUTE` statements in your `A10-LOOP` subroutine are performing calculations based on both variable types, it's no wonder there is a lot of converting from Binary to Decimal going on!

Try declaring variables X and Y as "COMP" too to see whether that helps improve performance any. I suspect it might.

**Comment by Fish-Git** [comment]
> My point is, try tweaking your COBOL program itself. It too seems to be inefficiently written IMHO. _<shrug>_

That having been said, I'm _**not**_ saying that Hercules's Decimal instructions implementation is as efficient as it can be. There may well be room for much improvement, I don't know. I have no investigated that angle yet.

All I'm saying is, asking the emulator to improve the inherent inefficiently of your COBOL programs is the wrong way to approach the problem. Instead, you should investigate how to write more efficient COBOL programs. It is not the emulator's responsibility to make up for poorly (inefficiently) written COBOL programs. Rather, it is the programmer's responsibility to write more efficient COBOL programs.

At least that's the way I'm seeing things right now.

As I said, I'm not saying there couldn't be some improvements that we could make to Hercules to improve the performance of our Decimal instructions. That may well be true.

But why should I/we go to that effort when there are obvious inherent inefficiencies in your program itself? As I said, it's not the responsibility of the emulator to make up for poorly written inefficient programs.

**Comment by wgs777** [comment]
I chose not to use COMP-2 for portability, as it's not supported by some non IBM compilers.
In MVS 3.8J a few minutes ago, I tried defining COMP-2 for X,Y, and ran the program and got all zeros in the "Y" result,
which is a different result than expected.

In referencing the IBM Cobol manual on page: 189, it details the rules for conversion of data types.  It appears to always do significant conversions no matter what data types are being used.  The manual also mentions that for longer data types over 15 digits like PIC 9(9)V9(9) it needs additional instructions and subroutine calls.  The performance hit for these conversions seems to be more severe than with other non IBM Cobol compilers.

Since we're only concerned here with Hercules, I've decided to temporarily take the Cobol compiler out of the equation.
Instead, I'll write a similar portable benchmark program in Fortran where Hercules should have the advantage
since the IBM Fortran H compiler is famous for its powerful optimizations.  I'll then run that program on the same emulators as before, and see if Hercules and the other emulators have similar timings.

Also, I'll do some additional testing on the IBM Cobol compiler (since I'll be needing to use it a lot on Hercules), using the most efficient data types to try to bring down the CPU usage.

If you do get time to take a very quick look at the coding for these instructions in Hercules: ZAP/PACK/UNPK/AP/SP,
just to see if there are any huge optimizations to be had it would be interesting to know.

Thanks again.

**Comment by wgs777** [comment]
I just wanted to provide an update.  I ran a few Fortran Benchmarks with Hercules and also my PDP-10 emulators.  Hercules was faster than my other emulators by a small margin with the Fortran benchmarks.
So, the performance issues I observed with Cobol were unrelated to Hercules.

In trying to tune the Cobol program with many strategic variable types, I've proven that this Cobol compiler generates 25 subroutine calls even when defining for X,Y variables for [1] COMP SYNC, [2] COMP-3, AND [3] COMP-2.
The run times for [1] and [2] above were both 50 seconds, matching the original "inefficient" run with X,Y being USAGE Display. 
For [3] COMP-2, the compiler is still generating assembler using these 25 subroutine calls.

  L     15,004(0,12)          V(ILBODTE1)
  L     15,008(0,12)          V(ILBODSP0)
  L     15,00C(0,12)          V(ILBOBID2)
  L     15,010(0,12)          V(ILBOIDB0)
  L     15,014(0,12)          V(ILBOBID1)
  L     15,010(0,12)          V(ILBOIDB0)
  L     15,00C(0,12)          V(ILBOBID2)
  L     15,00C(0,12)          V(ILBOBID2)
  L     15,010(0,12)          V(ILBOIDB0)
  L     15,014(0,12)          V(ILBOBID1)
  L     15,010(0,12)          V(ILBOIDB0)
  L     15,00C(0,12)          V(ILBOBID2)
  L     15,00C(0,12)          V(ILBOBID2)
  L     15,010(0,12)          V(ILBOIDB0)
  L     15,014(0,12)          V(ILBOBID1)
  L     15,010(0,12)          V(ILBOIDB0)
  L     15,00C(0,12)          V(ILBOBID2)
  L     15,004(0,12)          V(ILBODTE1)
  L     15,008(0,12)          V(ILBODSP0)
  L     15,008(0,12)          V(ILBODSP0)
  L     15,00C(0,12)          V(ILBOBID2)
  L     15,00C(0,12)          V(ILBOBID2)
  L     15,00C(0,12)          V(ILBOBID2)
  L     15,018(0,12)          V(ILBOXDI0)
  L     15,01C(0,12)          V(ILBOSTP1)

My thoughts here are that this compiler relies too much on subroutine calls.  This slowness wouldn't be noticable in a normal Cobol program, but becomes evident with a benchmark program.  Thanks all for responding.  I really appreciated your time on this.

**Comment by wgs777** [comment]
I saw some commits on the CLCLE and TRE instruction performance improvement.
Would anyone know if these 2 instruction's have performance improvements and if so, what % improvement?

**Comment by Fish-Git** [comment]
> I saw some commits on the `CLCLE` and `TRE` instruction performance improvement.
> 
> Would anyone know if these 2 instruction's have performance improvements and if so, what % improvement?

You can determine the improvement in performance yourself by running either one of the performance tests (`CLCLE-04-performance.tst` and `TRE-02-performance.tst`) in the "tests" subdirectory yourself on a version of Hercules both before the commit and then again afterwards.

If you need help on how to run the tests,, just ask. The most important thing you'll need to do to run them however, is to _enable_ them beforehand (they're both disabled by default) by uncommenting the appropriate statements in each test script. _(Refer the the test scripts themselves for more information.)_

Tests performed by both James (the author of the commits) and myself have measured improvements in the 95%+ range. A truly dramatic increase! For details, simply refer to the Git "Pull" requests themselves:

* [CLCLE instruction performance](https://github.com/SDL-Hercules-390/hyperion/pull/500)
* [TRE instruction performance](https://github.com/SDL-Hercules-390/hyperion/pull/498)

**Comment by wgs777** [comment]
Thank you.  I will try out the tests before getting the updates and after.

I have one unrelated question - not sure if this the correct place to ask
this question.
Apologies if I am posting it in the wrong forum or place.  Please let me
know the proper forum to ask it.

Here is the question:
I have hercules + gists fully built using gcc 12.2.1 and my hercules runs
Tk4- MVS 3.8J fine
with 2 emulated MVS CPU's.  Everything seems to work perfectly.
But I noticed in the startup logs, I get the following error:

HHC90020W 'hthread_setschedparam()' failed at loc=commadpt.c:1310: rc=22:
Invalid argument

The error occurs multiple lines (see below).  Would you know what the fix
would be.  I assume its a configuration issue?
Look for the words: Invalid argument at the end of several of the lines.

HHC01004I 0:066A COMM: listening on port 37913 for incoming TCP connections
HHC00100I Thread id 00007f2f749a7640, prio 0, name '3705 device(1:0669)
thread' started
HHC00100I Thread id 00007f2f747a5640, prio 0, name '3705 device(1:066A)
thread' started
HHC01004I 0:066B COMM: listening on port 37914 for incoming TCP connections
HHC00100I Thread id 00007f2f745a3640, prio 0, name '3705 device(1:066B)
thread' started
HHC90020W 'hthread_setschedparam()' failed at loc=commadpt.c:1310: rc=22:
Invalid argument
HHC00100I Thread id 00007f2f74493640, prio 0, name '0:0670 communication
thread' started
HHC01004I 0:0670 COMM: listening on port 37801 for incoming TCP connections
HHC90020W 'hthread_setschedparam()' failed at loc=commadpt.c:1310: rc=22:
Invalid argument
HHC00100I Thread id 00007f2f74392640, prio 0, name '0:0671 communication
thread' started
HHC01004I 0:0671 COMM: listening on port 37802 for incoming TCP connections
HHC90020W 'hthread_setschedparam()' failed at loc=commadpt.c:1310: rc=22:
Invalid argument
HHC00100I Thread id 00007f2f74291640, prio 0, name '0:0672 communication
thread' started

Thanks,
Bill

**Comment by Fish-Git** [comment]
Hi Bill!

> I have one unrelated question - not sure if this the correct place to ask this question. Apologies if I am posting it in the wrong forum or place. Please let me know the proper forum to ask it.

This is _**not**_ the correct forum/place. GitHub Issues are meant for reporting bugs or requesting enhancements or for Hercules developer discussions, etc. It is _not_ meant for providing help with running Hercules (e.g. simple user/usage  questions).

Additionally, we try to keep each GitHub Issue _focused_ on the problem/issue at hand, and try hard to not let the discussion stray from the original topic. All of Hercules's various support forums _(we have many!)_ are documented on our "Technical Support" web page:

* https://sdl-hercules-390.github.io/html/hercsupp.html

I would suggest you ask your question on the primary Hercules emulator support forum. Someone there should be able to help you. (I _could_ help you right here and now but am purposely refraining from doing so because I don't want to set a precedent by doing so.)

Hope that helps!

_(P.S. See below!)_
____

_**FYI:** I would very much appreciate it if you would _**not**_ respond/reply to GitHub Issues via email. **(\*)**_

_I would much prefer that you instead respond/reply _directly_ via the GitHub Issues web page itself:_

* **[ht<span>tps</span>://github.com/SDL-Hercules-390/hyperion/issues](https://github.com/SDL-Hercules-390/hyperion/issues)**

_When you reply _directly_ via their web page, I can make minor edits to your reply so it is more readable (prettier) by editing the fonts being used, formatting the log messages, etc._

_When you reply via email however, I cannot edit your reply (GitHub does not allow it), so oftentimes it is much harder (more difficult) to read._

_It is up to you whether or not you want to take the time to reply via their web page or continue to reply via email, but I would much rather that you reply directly via their web page instead._

_Thanks_

_**(\*)** GitHub does not support formatting of email replies, making it impossible for me to fix the formatting of a person's reply for readability. Thank you for understanding._

**Comment by Fish-Git** [comment]
Well, James Wekel has only just today managed to complete the one final remaining instruction listed in this GitHub Issue, so now between Bob Polmanter and James Wekel, _ALL_ of the instructions listed have had their performance improved _quite dramatically_ from how slow they were executing before! _THANK YOU, you guys!_ You'er the best! &nbsp;`:))`

Closing issue as completed! &nbsp;`:)`
